### PR TITLE
TELCODOCS-1277: first draft

### PR DESCRIPTION
**Pull request by StephenJamesSmith**
D/S Docs: [MGMT-13870](https://issues.redhat.com//browse/MGMT-13870) Orderly kmod upgrades

Jira: https://issues.redhat.com/browse/TELCODOCS-1277 (Upgrading kernel modules without a reboot)

Version(s): KMMO 1.1

Link to docs preview: https://63309--docspreview.netlify.app/openshift-enterprise/latest/hardware_enablement/kmm-kernel-module-management#kmm-customizing-upgrades-for-kernel-modules_kernel-module-management-operator

SME review: @yevgeny-shnaidman, @qbarrand 
QE review: @cdvultur

### CODE REVIEW

File: modules/kmm-upgrading-kernel-modules-without-a-reboot.adoc
Comment (qbarrand) [inline modules/kmm-upgrading-kernel-modules-without-a-reboot.adoc:12]: This feature is not specifically designed to upgrade a kernel module without a reboot. It was implemented to give users more control on the upgrade process, so that they can run any operation they want on the nodes - including a reboot, if needed.
For the title, I would go with Dan's suggestion "Customizing upgrades for kernel modules".

File: modules/kmm-upgrading-kernel-modules-without-a-reboot.adoc
Comment (yevgeny-shnaidman) [reply modules/kmm-upgrading-kernel-modules-without-a-reboot.adoc:104]: I agree, customer should not be aware of inner implementation.

Comment (StephenJamesSmith) [reply modules/kmm-upgrading-kernel-modules-without-a-reboot.adoc:104]: @yevgeny-shnaidman Are you suggesting removing the sections "Component implementation" and "Flow implementation"?

Comment (yevgeny-shnaidman) [reply modules/kmm-upgrading-kernel-modules-without-a-reboot.adoc:104]: yes, since it is internal implementation and should not concern users. If they want to know more, they can find it in the repo docs.

File: modules/kmm-upgrading-kernel-modules-without-a-reboot.adoc
Comment (StephenJamesSmith) [reply modules/kmm-upgrading-kernel-modules-without-a-reboot.adoc]: @qbarrand Changed the title and rewrote the intro paragraph.

File: modules/kmm-upgrading-kernel-modules-without-a-reboot.adoc
Comment (StephenJamesSmith) [reply modules/kmm-upgrading-kernel-modules-without-a-reboot.adoc]: @yevgeny-shnaidman Removed these sections.

**Comment by StephenJamesSmith** [comment]
@qbarrand If you have no more comments, please comment /LGTM. 
@cdvultur Please review and comment /LGTM if the content looks good to you.

File: modules/kmm-upgrading-kernel-modules-without-a-reboot.adoc
Comment (qbarrand) [inline modules/kmm-upgrading-kernel-modules-without-a-reboot.adoc:6]: Do we want to update that ID as well?

File: hardware_enablement/kmm-kernel-module-management.adoc
Comment (qbarrand) [inline hardware_enablement/kmm-kernel-module-management.adoc:69]: Do we want to change the filename as well?

File: modules/kmm-upgrading-kernel-modules-without-a-reboot.adoc
Comment (StephenJamesSmith) [reply modules/kmm-upgrading-kernel-modules-without-a-reboot.adoc]: Ah, right. changed.

File: hardware_enablement/kmm-kernel-module-management.adoc
Comment (StephenJamesSmith) [reply hardware_enablement/kmm-kernel-module-management.adoc]: Changed.

**Comment by StephenJamesSmith** [comment]
@cdvultur Please review and provide /LGTM if you approve the content. I'm waiting on your QE review before I can push to Peer Review. Thx!

**Comment by cdvultur** [comment]
/lgtm

Looks great!

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (aireilly) [inline modules/kmm-customizing-upgrades-for-kernel-modules.adoc:9]: Please remove the hard-wrapped lines. Update your editor preferences. See https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#general-file-guidelines

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (aireilly) [inline modules/kmm-customizing-upgrades-for-kernel-modules.adoc:7]: Is this title accurate? Are we customizing how to do an upgrade? What about just "Upgrading the kernel on a cluster node"

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (aireilly) [inline modules/kmm-customizing-upgrades-for-kernel-modules.adoc:9]: The sentence "Use this procedure to upgrade the kernel module while running maintenance operations on the node, including rebooting the node, if needed." is not very clear. 
```suggestion
Use this procedure to upgrade the kernel module while running maintenance operations on the node, including rebooting the node.
```

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (aireilly) [inline modules/kmm-customizing-upgrades-for-kernel-modules.adoc:10]:
```suggestion
To minimize the impact on workloads running in the cluster, run the kernel upgrade process sequentially, one node at a time.
```

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (aireilly) [inline modules/kmm-customizing-upgrades-for-kernel-modules.adoc:11]: Not sure this sentence makes sense: "Because it requires knowledge of the workload utilizing the kernel module, this process is managed by the cluster administrator."

I think you can safely remove it.

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (aireilly) [inline modules/kmm-customizing-upgrades-for-kernel-modules.adoc:16]: I think you can edit these prereqs for brevity. what about: 

```suggestion
* Terminate all user application workloads on the node or move them to another node.
* Unload the currently loaded kernel module. 
```

Also, are these actually prereqs, or should they be part of the procedure to upgrade the kernel on the node? If they are, I think you should provide concrete steps and commands to do these things.

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (aireilly) [inline modules/kmm-customizing-upgrades-for-kernel-modules.adoc:21]: I find it difficult to understand this prereq. I think it's not required in any case? Suggest removing it or clarifying what you mean. 

```suggestion
```

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (aireilly) [inline modules/kmm-customizing-upgrades-for-kernel-modules.adoc:25]: This doesn't seem like a prereq to me. IMO this should be a step in the procedure that needs to be documented.

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (aireilly) [inline modules/kmm-customizing-upgrades-for-kernel-modules.adoc:28]: This isn't a prereq - more like a note on the labeling step of the procedure.

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (aireilly) [inline modules/kmm-customizing-upgrades-for-kernel-modules.adoc:34]: Provide an `oc` command or an example that applies the label to the specified node.

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (aireilly) [inline modules/kmm-customizing-upgrades-for-kernel-modules.adoc:36]: Update the following fields of what CR?

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (aireilly) [inline modules/kmm-customizing-upgrades-for-kernel-modules.adoc:40]: I don't understand what atomic means in this step. Can you elaborate?

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (aireilly) [inline modules/kmm-customizing-upgrades-for-kernel-modules.adoc:42]: Suggest providing an example `oc` command or set of steps to do this. Also, what is kmod? First mention if this tool here.

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (aireilly) [inline modules/kmm-customizing-upgrades-for-kernel-modules.adoc:44]: Can we provide an `oc` command to do this, and a way to verify that the label was successfully removed?

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (aireilly) [inline modules/kmm-customizing-upgrades-for-kernel-modules.adoc:49]: Again, would like to see an `oc` command to do this, and some way to verify that the label was applied successfully.

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (aireilly) [inline modules/kmm-customizing-upgrades-for-kernel-modules.adoc:50]: What is `Module` in this context? Is it a CR? A YAML example might clarify what you are talking about.

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (aireilly) [inline modules/kmm-customizing-upgrades-for-kernel-modules.adoc:52]: Can you provide an example command or set of steps that shows how to do this?

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (aireilly) [inline modules/kmm-customizing-upgrades-for-kernel-modules.adoc:54]: This note could be incorporated into the procedure after step 4 above I think.

**Review by aireilly — CHANGES_REQUESTED** [review]
Hi Stephen, 

This looks good, but I think you need some additional detail and examples in the procedure steps. This reads quite differently to the procedure as outline in https://github.com/kubernetes-sigs/kernel-module-management/issues/263#issue-1553393178. Also the Prereqs could be clarified and reduced somewhat. Reach out if you want to discuss. Thanks.

**Comment by StephenJamesSmith** [comment]
@aireilly Thanks for the review, Aidan. Yes, I would like to discuss your comment about using https://github.com/kubernetes-sigs/kernel-module-management/issues/263#issue-1553393178. This is not mentioned in the Jira and the SME and QE have both signed off on this PR.  I'll arrange a meeting for thi Monday. Have a great recharge!

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (StephenJamesSmith) [reply modules/kmm-customizing-upgrades-for-kernel-modules.adoc:9]: Removed.

Comment (StephenJamesSmith) [reply modules/kmm-customizing-upgrades-for-kernel-modules.adoc:9]: s/including/ such as

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (StephenJamesSmith) [reply modules/kmm-customizing-upgrades-for-kernel-modules.adoc:11]: Changed to "This procedure requires knowledge of the workload utilizing the kernel module and must be managed by the cluster administrator."

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (StephenJamesSmith) [reply modules/kmm-customizing-upgrades-for-kernel-modules.adoc:16]: Made these changes.

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (StephenJamesSmith) [reply modules/kmm-customizing-upgrades-for-kernel-modules.adoc:21]: No change.

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (StephenJamesSmith) [reply modules/kmm-customizing-upgrades-for-kernel-modules.adoc:28]: Will address this in the enhancement story.

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (StephenJamesSmith) [reply modules/kmm-customizing-upgrades-for-kernel-modules.adoc:34]: Will address this in the enhancement story.

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (StephenJamesSmith) [reply modules/kmm-customizing-upgrades-for-kernel-modules.adoc:36]: "Update the following fields in the `Module` CR"

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (StephenJamesSmith) [reply modules/kmm-customizing-upgrades-for-kernel-modules.adoc:42]: Will address this in the enhancement story.

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (StephenJamesSmith) [reply modules/kmm-customizing-upgrades-for-kernel-modules.adoc:44]: Will address this in the enhancement story.

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (StephenJamesSmith) [reply modules/kmm-customizing-upgrades-for-kernel-modules.adoc:49]: Will address this in the enhancement story.

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (StephenJamesSmith) [reply modules/kmm-customizing-upgrades-for-kernel-modules.adoc:50]: Added "CR".

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (StephenJamesSmith) [reply modules/kmm-customizing-upgrades-for-kernel-modules.adoc:52]: Will address this in the enhancement story.

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (StephenJamesSmith) [reply modules/kmm-customizing-upgrades-for-kernel-modules.adoc:54]: No change for now.

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (StephenJamesSmith) [reply modules/kmm-customizing-upgrades-for-kernel-modules.adoc]: Will address this in the enhancement story.

**Comment by StephenJamesSmith** [comment]
@aireilly Do you need to review again before I add a /label merge-review-needed?

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (aireilly) [inline modules/kmm-customizing-upgrades-for-kernel-modules.adoc:51]: I still feel this note is confusing tbh.

Comment (StephenJamesSmith) [reply modules/kmm-customizing-upgrades-for-kernel-modules.adoc:51]: @Aidan I spoke with Costi and we came up this - "If no additional upgrading is needed, you can skip Steps 3 through 6 by updating the `[kmm.node.kubernetes.io/version-module](http://kmm.node.kubernetes.io/version-module).<module-namespace>.<module-name>` label value to the new `$moduleVersion` as set in the `Module`."

Comment (aireilly) [reply modules/kmm-customizing-upgrades-for-kernel-modules.adoc:51]: OK cool. I presume what we are getting at here is that you can only upgrade a single kernel module at a time. Is this the case?

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (aireilly) [inline modules/kmm-customizing-upgrades-for-kernel-modules.adoc:31]: So is this actually a prereq to this procedure? "Before upgrading..."

Comment (StephenJamesSmith) [reply modules/kmm-customizing-upgrades-for-kernel-modules.adoc:31]: I'll make this a prereq.

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (StephenJamesSmith) [reply modules/kmm-customizing-upgrades-for-kernel-modules.adoc:51]: @aireilly From Costi - "you can do multiple at the same time, if you respect the procedure"

Comment (StephenJamesSmith) [reply modules/kmm-customizing-upgrades-for-kernel-modules.adoc]: /label merge-review-needed

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (stevsmit) [inline modules/kmm-customizing-upgrades-for-kernel-modules.adoc:30]: It seems weird to me that there are two sets of prerequisites. Also, these prerequisites seems like an entire procedure? Which has me thinking that maybe this instance of the word .Prerequisites should be deleted?

File: modules/kmm-customizing-upgrades-for-kernel-modules.adoc
Comment (stevsmit) [inline modules/kmm-customizing-upgrades-for-kernel-modules.adoc:33]: Should this be a prerequisite step, and the actual procedure begin with "Update the following. . ." ?

**Comment by StephenJamesSmith** [comment]
@stevsmit That is correct. We had a lot of discussion about this. The Devs and QE feel there is enough information in this story to be useful to the user, albeit written in a "broad strokes" style. However, some reviewers thought it could use more info, such as commands and examples for certain sections. The agreement was to release this and then create an Enhancement story to have a second look as this technology evolves. So, yes - please merge.

**Comment by stevsmit** [comment]
@StephenJamesSmith I'm happy to merge this but am requesting one change: Will you fix the double ".Prerequisite" sections and the ".Procedure" heading so that the actual procedure is underneath it? Please + thanks. After that, I'll be happy to merge it. It seems like a trivial change but I brought this up with my peer review squad and they requested that, at the very least, those sections were fixed for this draft. 

Thanks! 

image

**Pull request merged**
